### DISCUSSION TITLE
How to use librosa and python_speech_features with Pyscript?

### DISCUSSION THREAD

**Discussion by rawhuul** [pyscript/pyscript#1348] [Q&A]
I'm working on a project using `Pyscript` and need to use some audio processing libraries such as `librosa` and `python_speech_features`. However, when I try to import them, I get an error message saying that these libraries can't be used. I know these libraries are not implemented in pure Python and require C libraries.

Is there a way to use these libraries with `Pyscript`? Are there any workarounds or alternatives that I can use instead?

Thanks for your help!

**Comment by hoodmane** [comment]
The `python_speech_features`  is pure Python package but its maintainers didn’t upload a wheel. You can [build a wheel](https://packaging.python.org/en/latest/tutorials/packaging-projects/#generating-distribution-archives) yourself, serve it, and install it with micropip from the corresponding URL.

librosa looks like it's easy enough to build:
https://github.com/librosa/librosa#building-from-source
You can try following the instructions here:
https://pyodide.org/en/stable/development/building-and-testing-packages.html

It requires `python-soundfile` which is a pure Python module that uses `libsndfile.so` via ctypes. libsndfile also sounds possible to build. Based on the instructions here: 
https://github.com/libsndfile/libsndfile#hacking
it can probably be built with:
```sh
autoreconf -vif
emconfigure ./configure --enable-werror
emmake make
```
though then you have to figure out how to get the resulting shared library into Pyodide. You can make a recipe (see the one for [libmagic.so](https://github.com/pyodide/pyodide/blob/main/packages/libmagic/meta.yaml) for comparison). Pyodide has out of tree recipe build support but I don't know if it works with shared libraries (@ryanking13?)

librosa also uses audioread which is a pure Python project which says [it has various backends](https://github.com/beetbox/audioread#audioread). Hopefully you can get away with the backends that are Python stdlib modules because otherwise you would need to build more stuff. Most of the backends sound like they can't be ported, but there are some candidates.

**Reply by rawhuul** [reply]
I will look into it, thanks <3

**Comment by rawhuul** [comment]
Hey there I was able to build wheel package for `python_speech_features`, how can I install it using micropip?

```
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <link rel="stylesheet" href="https://pyscript.net/latest/pyscript.css" />
    <script defer src="https://pyscript.net/latest/pyscript.js"></script>

</head>

<body>
    <py-config>
        packages = [ "numpy", "scikit-learn", "scipy"]
    </py-config>

    <py-script>
        import micropip
        import asyncio

        async def install_speech():
            await micropip.install("emfs://assets/python_speech_features-0.6.1-py3-none-any.whl")

        asyncio.ensure_future(install_speech())

        import python_speech_features as ps
        print(ps)
    </py-script>
</body>

</html>
```

But it results in:
![image](https://user-images.githubusercontent.com/84739019/229344402-3c7431a1-a313-41cc-83df-3716c8b1b957.png)

It seems like I'm doing something wrong, how can I fix it?

**Reply by hoodmane** [reply]
Two problems:
1. `ensure_future` schedules a task but doesn't do it so you can't just import the module after that 
2. `emfs:` is for installing the wheel from the Emscripten file system. You need to serve the wheel and install from a URL. Or if using chrome you could mount the native fs and use an emfs URL but serving the wheel is more normal.

**Reply by rawhuul** [reply]
Can you please explain it in details, I never done this before. A little example of how to ship a pre-built package and using it in `py-script` would be much helpful.

**Reply by JeffersGlass** [reply]
For Hood's point 2: You'll need some basic way of making the wheel available 'on the network'. If you're developing locally, a [simple development server](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/Tools_and_setup/set_up_a_local_testing_server#running_a_simple_local_http_server) will work fine. If you're hosting the project on the web somewhere, you're probably good to go. Either way, what you need is the URL of the wheel file, either on the web or a local address. 

You'll want to add the URL for the wheel to your list of `packages` to install in \<py-config\>. So long as the URL ends in `.whl`, Micropip will know that it's a wheel (and not, say, a package from PyPI or Pyodide) and handle installing it. (The `packages` key is a pretty thin wrapper around Micropip, at this point, for what it's worth.)

This also speaks to Hood's point 1 - `asyncio.ensure_future` schedules a task to run asynchronously, but doesn't guarentee it completes before your code moves along to the next line. Which is to say, in your code example, the wheel will probably not have finished installing by the time your code hits `import python_speech_features as ps`, and so the import fails.

That said, when you install via the `packages` key, PyScript will make sure all the packages have finished installing before your code runs, so you could be able to just `import python_speech_features as ps` in your code.

**Reply by rawhuul** [reply]
I tried using it as package in `py-config`. But it still returns an error.

This is my code:
![image](https://user-images.githubusercontent.com/84739019/232127377-da84e7a4-fa3d-4a7d-a522-edc5a7ce857a.png)

It is result using `http.server` from python:
![image](https://user-images.githubusercontent.com/84739019/232127476-50522a58-fa05-4251-aab0-7df76285b355.png)

It is result using `parcel`:
![image](https://user-images.githubusercontent.com/84739019/232129545-56486561-e972-41a2-aea4-f33cabb86009.png)

**Reply by rawhuul** [reply]
I got it working by simply hosting a file on my github repo. Thanks for help <3

**Comment by ryanking13** [comment]
> Pyodide has out of tree recipe build support but I don't know if it works with shared libraries (@ryanking13?)

It works with shared libraries too. But I'm not sure micropip will locate the path to shared libraries correctly. I'll check when I have bandwidth.
